### PR TITLE
New assign color options

### PR DESCRIPTION
**Pull request by frheault**
To facilitate visualization in MI-Brain, allow scil_assign_color_to_tractogram.py to color using a nifti to get coloring information.
Also support a LUT to map labels to values (such as pvalue along the bundle).
image

image

### CODE REVIEW

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by frheault** [comment]
@CHrlS98 Can you take a look at the new scripts for the options?

I will add a resampling option tomorrow. Or do you think I should simply detect large step-size and warn the user to use our other script to do that?

**Comment by CHrlS98** [comment]
I'll do a proper review later, but for your second question it makes sense to just warn the user if the number of points is smaller than the number of unique voxel labels intersected by the streamline!

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_assign_uniform_color_to_tractogram.py [hunk 1]
```diff
@@ -38,25 +35,14 @@ def _build_arg_parser():
     p.add_argument('in_tractograms', nargs='+',
```
Comment (CHrlS98) [inline scripts/scil_assign_uniform_color_to_tractogram.py:35]: Should the script name be `..._to_tractograms.py` since it can take more than 1 input?

**Review by CHrlS98 — COMMENTED** [review]
Cool!

For the `resampling` part, the warning could reference the script `scil_resample_streamlines.py`.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    g1 = p.add_argument_group(title='Coloring Methods')
+    p1 = g1.add_mutually_exclusive_group()
+    p1.add_argument('--use_dps', metavar='DPS_KEY',
+                    help='Use the data_per_streamline (scalar) for coloring,\n'
+                         'linear scaling from min-max, e.g. commit_weights.')
+    p1.add_argument('--use_dpp', metavar='DPP_KEY',
+                    help='Use the data_per_point (scalar) for coloring,\n'
+                         'linear scaling from min-max.')
+    p1.add_argument('--load_dps', metavar='DPS_KEY',
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_assign_custom_color_to_tractogram.py:64]: I added it to the description.

**Comment by CHrlS98** [comment]
@frheault I tested the PR.
I tried running the command
```
>>> scil_assign_custom_color_to_tractogram.py sub-1108__228_246_uniformized.trk \
... colored_pval.trk --from_anatomy voxmap_100.nii.gz --LUT 228_246_stats_pc2.txt
```
with my data ([here](https://usherbrooke-my.sharepoint.com/:f:/g/personal/poic2312_usherbrooke_ca/Evk0ToVEuPxPsGzVZI7sN34BaAzDaCw46ZxmGBR4VCciOA?e=yuMd57)), but I can't seem to get the right result! My bundle is just all blue.

image

Am I doing something wrong or is there a bug? There are 100 labels and 100 values in my LUT. My streamlines do not all have 100 points, but I tried with resampled streamlines and still get the same result.

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 3]
```diff
@@ -93,18 +93,18 @@ def _build_arg_parser():
 
 def transform_data(args, data):
     if args.LUT:
-        data = np.round(data).astype(np.int32)
         LUT = load_matrix_in_any_format(args.LUT)
+        print(LUT)
```
Comment (CHrlS98) [inline scripts/scil_assign_custom_color_to_tractogram.py:97]: debug trace

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 4]
```diff
@@ -93,18 +93,18 @@ def _build_arg_parser():
...
         for i, val in enumerate(LUT):
             data[data == i+1] = val
+            print(i+1, val)
```
Comment (CHrlS98) [inline scripts/scil_assign_custom_color_to_tractogram.py:100]: here too

**Review by CHrlS98 — COMMENTED** [review]
Cool it works, once the debug trace is removed I think it will be ready to merge!

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_compute_bundle_voxel_label_map.py [hunk 5]
```diff
@@ -49,10 +49,10 @@ def _build_arg_parser():
     p.add_argument('--nb_pts', type=int,
                    help='Number of divisions for the bundles.\n'
                         'Default is the number of points of the centroid.')
-    p.add_argument('--min_streamline_count', type=int, default=100,
+    p.add_argument('--min_streamline_count', type=int, default=100000,
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_compute_bundle_voxel_label_map.py:52]: That's a question for @mdesco , I see now that virtually everyone using scil_compute_bundle_voxel_label_map.py does not save the TRK (optional, but internally the TRK has been trimmed), but then use the voxel labels map with the original TRK (not trimmed) and there is a discrepancy.

This trimming seems to increase risk of error, since most people don't know its happening. This could be problematic when used outside of tractometry flow. So what do we do?

File: scripts/scil_assign_uniform_color_to_tractograms.py [hunk 6]
```diff
@@ -95,9 +83,8 @@ def main():
         out_filenames.append('{}{}{}'.format(base, args.out_suffix, '.trk'))
     assert_outputs_exist(parser, args, out_filenames)
 
-    sft = load_tractogram_with_reference(parser, args, filename)
-    cmap = plt.get_cmap(args.colormap)
     for i, filename in enumerate(args.in_tractograms):
+        sft = load_tractogram_with_reference(parser, args, filename)
```
Comment (grahamlittlephd) [inline scripts/scil_assign_uniform_color_to_tractograms.py:87]: @arnaudbore should probably merge this into the scilus container ASAP.  Without moving this line in the for loop, tractometry_flow just outputs a bunch of Right UF tracks in the Color_Bundles folder with different colors.

File: scripts/scil_compute_bundle_voxel_label_map.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (CHrlS98) [reply scripts/scil_compute_bundle_voxel_label_map.py:52]: Ok I think I get it. But if we really want to make sure we don't trim anything shouldn't we set the default to `None` instead of `100000`?

**Review by mdesco — COMMENTED** [review]
Can you guys prepare a small dataset for me to test this PR well please? And maybe a run.sh command to see a bit better.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+A minimum and a maximum range can be provided to clip values. If the range of
+values is too large for intuitive visualization, a log transform can be applied.
+
+If the data provided from --use_dps, --use_dpp and --from_anatomy are integer
+labels, they can be mapped using a LookUp Table (--LUT).
+The file provided as a LUT should be either .txt or .npy and if the
+size is N=20, then the data provided should be between 1-20.
+
+Example: Use --anatomy with a voxel labels map (values from 1-20) with a text
+file containing 20 p-values to map p-values to the bundle for visualisation.
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_assign_custom_color_to_tractogram.py:26]: Once you test it and leave comments I will do all of it at once.

**Comment by CHrlS98** [comment]
@mdesco 
For p-value mapping, you can run
```
>>> scil_assign_custom_color_to_tractogram.py sub-1108__228_246_uniformized.trk \
... colored_pval.trk --from_anatomy voxmap_100.nii.gz --LUT 228_246_stats_pc2.txt
```
with the data [here](https://usherbrooke-my.sharepoint.com/:f:/g/personal/poic2312_usherbrooke_ca/Evk0ToVEuPxPsGzVZI7sN34BaAzDaCw46ZxmGBR4VCciOA?e=yuMd57). @frheault might have more data for other use cases.

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+The script use scalar in data_per_point and data_per_streamline
```
Comment (mdesco) [inline scripts/scil_assign_custom_color_to_tractogram.py:5]: The script uses the scalar

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+(e.g commit_weights) to visualise on the streamlines.
```
Comment (mdesco) [inline scripts/scil_assign_custom_color_to_tractogram.py:6]: to visualize them on the

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+Saves the RGB values in the data_per_point (color_x, color_y, color_z).
+
+If called with .tck, the output will always be .trk, because data_per_point has
+no equivalent in tck file.
+
+The usage of --use_dps, --use_dpp and --from_anatomy is more complex. It maps
+the raw values from these sources to RGB using a colormap.
+    --use_dps: nbr. streamlines = len(streamlines)
+    --use_dpp: nbr. points = len(streamlines._data)
```
Comment (mdesco) [inline scripts/scil_assign_custom_color_to_tractogram.py:15]: I don't understand this. First, please avoid to use the '.' 
You mean number of streamlines = len(streamlines) ?
number of points ?
streamlines._data -> what does that mean?

File: scripts/scil_compute_bundle_voxel_label_map.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdesco) [reply scripts/scil_compute_bundle_voxel_label_map.py:52]: @frheault I agree with your EDIT idea above. 

Also missing space between cutting and operations

File: scripts/scil_compute_bundle_voxel_label_map.py [hunk 11]
```diff
@@ -49,10 +49,10 @@ def _build_arg_parser():
...
                    help='Minimum number of streamlines for filtering/cutting'
                         'operation [%(default)s].')
-    p.add_argument('--min_voxel_count', type=int, default=1000,
+    p.add_argument('--min_voxel_count', type=int, default=1000000,
                    help='Minimum number of voxels for filtering/cutting'
                         'operation [%(default)s].')
```
Comment (mdesco) [inline scripts/scil_compute_bundle_voxel_label_map.py:57]: missing space between cutting and operations

**Review by mdesco — APPROVED** [review]
1. Minor edits to resolve in the help
2. Otherwise, here a few examples of command line I tried that could serve the help. It would be nice to see an example with dpp and dps... 
scil_assign_custom_color_to_tractogram.py sub-1108__228_246_uniformized.tck  --reference voxmap_100.nii.gz --from_anatomy voxmap_100.nii.gz --LUT 228_246_stats_pc2.txt test_pvalues.trk

scil_assign_custom_color_to_tractogram.py sub-1108__228_246_uniformized.tck --from_anatomy voxmap_100.nii.gz test.trk --reference voxmap_100.nii.gz

scil_assign_custom_color_to_tractogram.py sub-1108__228_246_uniformized.tck --from_anatomy voxmap_100.nii.gz --reference voxmap_100.nii.gz --log test2.trk

**Comment by EmmaRenauld** [comment]
Just leaving this here. I was thinking the other day that I would have really liked to have a custom color where (close to) the seeding point would be a chosen color, and the rest, uniform. Could we eventually use your script to get something like that directly from the data_per_streamline? I managed by visualising the seed density map instead, but it's less precise!

**Comment by frheault** [comment]
> Just leaving this here. I was thinking the other day that I would have really liked to have a custom color where (close to) the seeding point would be a chosen color, and the rest, uniform. Could we eventually use your script to get something like that directly from the data_per_streamline? I managed by visualising the seed density map instead, but it's less precise!

An option could be added just for that. But it would require a for loop to find the closest point on the streamline for the seed (pretty slow).

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_clean_qbx_clusters.py [hunk 12]
```diff
@@ -96,7 +96,7 @@ def keypress_callback(obj, _):
             return
 
         if key == 'q':
-            show_manager.exit()
+            #show_manager.exit()
```
Comment (arnaudbore) [inline scripts/scil_clean_qbx_clusters.py:99]: @frheault do you want to keep this ?

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-566/8/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-566/9/

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_clean_qbx_clusters.py [hunk 13]
```diff
@@ -240,9 +242,8 @@ def keypress_callback(obj, _):
                                          args.out_accepted_dir,
                                          filename_accepted_on_size)
 
-    accepted_sft = StatefulTractogram(accepted_streamlines,
-                                      sft_accepted_on_size[0],
-                                      Space.RASMM)
+    accepted_sft = StatefulTractogram.from_sft(accepted_streamlines,
```
Comment (CHrlS98) [inline scripts/scil_clean_qbx_clusters.py:245]: Why are these changes included?

File: scripts/scil_clean_qbx_clusters.py [hunk 14]
```diff
@@ -257,9 +258,8 @@ def keypress_callback(obj, _):
                                               args.out_rejected_dir,
                                               filename_rejected_on_size))
 
-    rejected_sft = StatefulTractogram(rejected_streamlines,
-                                      sft_accepted_on_size[0],
-                                      Space.RASMM)
+    rejected_sft = StatefulTractogram.from_sft(rejected_streamlines,
```
Comment (CHrlS98) [inline scripts/scil_clean_qbx_clusters.py:261]: Same comment here?

**Review by CHrlS98 — COMMENTED** [review]
@frheault I don't understand why changes to bundle_voxel_label_map are included in this PR?

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+The script uses scalars in data_per_point and data_per_streamline
```
Comment (CHrlS98) [inline scripts/scil_assign_custom_color_to_tractogram.py:5]: or from anatomical  image?

File: scripts/scil_assign_custom_color_to_tractogram.py [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+A minimum and a maximum range can be provided to clip values. If the range of
+values is too large for intuitive visualization, a log transform can be applied.
+
+If the data provided from --use_dps, --use_dpp and --from_anatomy are integer
+labels, they can be mapped using a LookUp Table (--LUT).
+The file provided as a LUT should be either .txt or .npy and if the
+size is N=20, then the data provided should be between 1-20.
+
+Example: Use --anatomy with a voxel labels map (values from 1-20) with a text
```
Comment (CHrlS98) [inline scripts/scil_assign_custom_color_to_tractogram.py:25]: `--from_anatomy`

**Review by CHrlS98 — COMMENTED** [review]
Minor comments on the documentation but otherwise good to go.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Review by CHrlS98 — APPROVED** [review]
Rock n roll on my end!
